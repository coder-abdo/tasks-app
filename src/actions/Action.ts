import React from "react";
import axios from "axios";
export const FETCH_DATA: string = "FETCH_DATA";
export const AUTHORIZED_USER: string = "AUTHORIZED_USER";
export const ADD_TASK: string = "ADD_TASK";
export const REMOVE_TASK: string = "REMOVE_TASK";
export const fetchingData = () => {
  return (dispatch: React.Dispatch<any>) => {
    axios.get("./data.json").then(({ data }) =>
      dispatch({
        type: FETCH_DATA,
        payload: {
          user: data.user,
          tasks: data.tasks
        }
      })
    );
  };
};
