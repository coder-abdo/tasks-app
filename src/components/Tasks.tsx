import React from "react";
import styled from "styled-components";
const AppTasks = styled.div`
  padding: 12vh 4vw 2vh;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 2vw;
  height: 100vh;
  @media (max-width: 600px) {
    grid-template-columns: 1fr;
    justify-items: center;
  }
`;
const Tasks = ({ children }) => {
  return <AppTasks>{children}</AppTasks>;
};

export default Tasks;
