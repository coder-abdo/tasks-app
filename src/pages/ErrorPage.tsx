import React from "react";
import { ErrorContainer } from "../styles/ErrorPageStyle";
export const ErrorPage = () => {
  return (
    <ErrorContainer>
      <h1>
        <span className="large">404</span>
        Ooops you lost
      </h1>
    </ErrorContainer>
  );
};
