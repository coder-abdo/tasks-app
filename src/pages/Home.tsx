import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { fetchingData, REMOVE_TASK } from "../actions/Action";
import Task from "../components/Task";
export const Home = () => {
  const tasks = useSelector((state: any) => state.tasks);
  const user = useSelector((state: any) => state.user);

  const dispatch = useDispatch();
  const getData = () => dispatch(fetchingData());
  useEffect(() => {
    getData();
  }, []);
  const handleRemoveTask = task =>
    dispatch({ type: REMOVE_TASK, payload: task });
  return (
    <div>
      {tasks.map((task: any) => (
        <Task
          key={task.id}
          {...task}
          handleRemoveTask={() => handleRemoveTask(task.id)}
        />
      ))}
    </div>
  );
};
