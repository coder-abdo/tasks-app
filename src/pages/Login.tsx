import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { useHistory } from "react-router-dom";
import { AUTHORIZED_USER } from "../actions/Action";
import { LoginContainer, LoginForm } from "../styles/LoginStyle";
import { IState } from "../interfaces/interfaces";

export const Login = () => {
  const userInfo = useSelector((state: IState) => state.user);
  const dispatch = useDispatch();
  const [username, setUsername] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [err, setErr] = useState<string>("");
  const history = useHistory();

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    if (username === userInfo.username && password === userInfo.password) {
      dispatch({ type: AUTHORIZED_USER });
      history.push("/");
    } else {
      if (username !== userInfo.username) {
        setErr("username does not matched");
      } else if (password !== userInfo.password) {
        setErr("password does not matched");
      } else {
        setErr("your password or username does not matched");
      }
    }
    setUsername("");
    setPassword("");
  };
  return (
    <LoginContainer>
      <h2>hello welcome back</h2>
      <LoginForm onSubmit={handleSubmit}>
        <input
          type="text"
          name="user"
          id="user"
          placeholder="Username"
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setUsername(e.target.value)
          }
          value={username}
          required
        />
        <input
          type="password"
          name="password"
          id="pass"
          placeholder="Password"
          value={password}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setPassword(e.target.value)
          }
          required
        />
        <button type="submit">Login</button>
      </LoginForm>
      {err && <div className="err-message">{err}</div>}
    </LoginContainer>
  );
};
