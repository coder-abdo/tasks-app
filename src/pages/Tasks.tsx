import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { ADD_TASK, REMOVE_TASK } from "../actions/Action";
import AppTasks from "../components/Tasks";
import Task from "../components/Task";
import TaskForm from "../styles/TasksStyle";
import { IState, ITask } from "../interfaces/interfaces";
export const Tasks = () => {
  const tasks: ITask[] = useSelector((state: IState) => state.tasks);
  const dispatch = useDispatch();
  const [title, setTitle] = useState("");
  const [cateogry, setCateogry] = useState("");
  const validateForm = () => title.length && cateogry.length;
  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (validateForm()) {
      dispatch({
        type: ADD_TASK,
        payload: {
          title,
          cateogry
        }
      });
    } else {
      alert("error");
    }

    setTitle("");
    setTitle("");
  };

  return (
    <div>
      <h2>control your tasks</h2>
      <TaskForm onSubmit={handleSubmit}>
        <input
          type="text"
          value={title}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setTitle(e.target.value)
          }
          placeholder="task title"
          required
        />
        <input
          type="text"
          value={cateogry}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setCateogry(e.target.value)
          }
          placeholder="task category"
          required
        />
        <button type="submit">Add Task</button>
      </TaskForm>
      <AppTasks>
        {tasks.map((task: ITask) => (
          <Task
            key={task.id}
            {...task}
            handleRemoveTask={() =>
              dispatch({ type: REMOVE_TASK, payload: task.id })
            }
          />
        ))}
      </AppTasks>
    </div>
  );
};
