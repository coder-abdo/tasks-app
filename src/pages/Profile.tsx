import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { FaRegUserCircle } from "react-icons/fa";
import { REMOVE_TASK } from "../actions/Action";
import Task from "../components/Task";
import Tasks from "../components/Tasks";
import { ProfileInfo } from "../styles/ProfileStyle";
import { IState, ITask } from "../interfaces/interfaces";
export const Profile = () => {
  const tasks: ITask[] = useSelector((state: IState) => state.tasks);
  const dispatch = useDispatch();
  return (
    <>
      <ProfileInfo>
        <FaRegUserCircle />
        <span>Admin </span>
        <span className="tasks-num">tasks: {tasks.length}</span>
      </ProfileInfo>
      <Tasks>
        {tasks.map((task: ITask) => (
          <Task
            key={task.id}
            {...task}
            handleRemoveTask={() =>
              dispatch({ type: REMOVE_TASK, payload: task.id })
            }
          />
        ))}
      </Tasks>
    </>
  );
};
