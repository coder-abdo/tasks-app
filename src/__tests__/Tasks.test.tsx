import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { Tasks } from "../pages/Tasks";
import { Provider } from "react-redux";
import Store from "../store/store";
import { ADD_TASK } from "../actions/Action";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";

beforeEach(cleanup);
describe("render Tasks page", () => {
  const store = configureStore({
    tasks: []
  });
  const addTask = jest.fn();
  it("should render the form of Tasks ", () => {
    const TaskPage = render(
      <Provider store={Store}>
        <Router>
          <Tasks />
        </Router>
      </Provider>
    );
    expect(TaskPage).toBeTruthy();
    expect(TaskPage).toMatchSnapshot();
  });
  it("should change the task title and task cateogry properly", () => {
    const { getByTestId } = render(
      <Provider store={Store}>
        <Router>
          <Tasks />
        </Router>
      </Provider>
    );
    expect(getByTestId("task-title")).toHaveValue("");
    expect(getByTestId("task-cateogry")).toHaveValue("");
    fireEvent.change(getByTestId("task-title"), {
      target: { value: "do laundry" }
    });
    fireEvent.change(getByTestId("task-cateogry"), {
      target: { value: "simpleTodo" }
    });
    expect(getByTestId("task-title")).toHaveValue("do laundry");
    expect(getByTestId("task-cateogry")).toHaveValue("simpleTodo");
    fireEvent.click(getByTestId("submiting"), { preventDefault() {} });
    expect(addTask.mock.calls.length).toEqual(0);
  });
});
