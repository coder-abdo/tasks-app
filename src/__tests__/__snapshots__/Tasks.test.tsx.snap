// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Tasks page should render the form of Tasks  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2>
          control your tasks
        </h2>
        <form
          class="sc-htpNat dAJjhG"
        >
          <input
            data-testid="task-title"
            placeholder="task title"
            required=""
            type="text"
            value=""
          />
          <input
            data-testid="task-cateogry"
            placeholder="task category"
            required=""
            type="text"
            value=""
          />
          <button
            data-testid="submiting"
            type="submit"
          >
            Add Task
          </button>
        </form>
        <div
          class="sc-bdVaJa kqCUAT"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2>
        control your tasks
      </h2>
      <form
        class="sc-htpNat dAJjhG"
      >
        <input
          data-testid="task-title"
          placeholder="task title"
          required=""
          type="text"
          value=""
        />
        <input
          data-testid="task-cateogry"
          placeholder="task category"
          required=""
          type="text"
          value=""
        />
        <button
          data-testid="submiting"
          type="submit"
        >
          Add Task
        </button>
      </form>
      <div
        class="sc-bdVaJa kqCUAT"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
